modules/programs/ncmpcpp.nix

```diff
@@ -0,0 +1,142 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) types mkIf mkEnableOption mkOption;
+
+  cfg = config.programs.ncmpcpp;
+
+  mpdCfg = config.services.mpd;
+
+  mpdMusicDir = if cfg.mpdMusicDir == null then
+    mpdCfg.musicDirectory
+  else
+    toString cfg.mpdMusicDir;
+
+  renderSettings = settings:
+    lib.concatStringsSep "\n" (lib.mapAttrsToList renderSetting settings);
+
+  renderSetting = name: value: "${name}=${renderValue value}";
+
+  renderValue = option:
+    rec {
+      int = toString option;
+      bool = if option then "yes" else "no";
+      string = option;
+    }.${builtins.typeOf option};
+
+  renderBindings = bindings:
+    lib.concatStringsSep "\n" (map renderBinding bindings);
+
+  renderBinding = { key, command }:
+    lib.concatStringsSep "\n  "
+    ([ ''def_key "${key}"'' ] ++ maybeWrapList command);
+
+  maybeWrapList = xs: if lib.isList xs then xs else [ xs ];
+
+  valueType = types.oneOf [ types.bool types.int types.str ];
+
+  bindingType = types.submodule ({ name, config, ... }: {
+    options = {
+      key = mkOption {
+        type = types.str;
+        description = "Key to bind.";
+        example = "j";
+      };
+
+      command = mkOption {
+        type = types.either types.str (types.listOf types.str);
+        description = "Command or sequence of commands to be executed.";
+        example = "scroll_down";
+      };
+    };
+  });
+
+in {
+  meta.maintainers = with lib.maintainers; [ olmokramer ];
+
+  options.programs.ncmpcpp = {
+    enable =
+      mkEnableOption "ncmpcpp - An ncurses Music Player Daemon (MPD) client";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.ncmpcpp;
+      defaultText = "pkgs.ncmpcpp";
+      description = "Package providing the <code>ncmpcpp</code> command.";
+      example = "pkgs.ncmpcpp.override { ... }";
+    };
+
+    mpdMusicDir = mkOption {
```

Comment:
We could get rid of this option and place it under `settings` if `settings` were a [freeform module](https://nixos.org/manual/nixos/unstable/#sec-freeform-modules), but that breaks the tests with an error:

```
Module `:anon-1:anon-1' has an unsupported attribute `freeformType'.
This is caused by introducing a top-level `config' or `options'
attribute. Add configuration attributes immediately on the top level
instead, or move all of them (namely: freeformType) into the
explicit `config' attribute.
```

The [Freeform modules PR](https://github.com/NixOS/nixpkgs/pull/82743) has only been merged 12 days ago, though, so it may not be a great idea to start relying on it already.

Reply:
This is what that might look like https://github.com/olmokramer/home-manager/blob/ncmpcpp-freeform/modules/programs/ncmpcpp.nix

Reply:
Cool. I'll have to check out 🙂